HIGH: (Comment linking: step 2) [23220] WEB maintain visible content position

perunt opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
 
This PR introduces the maintainVisibleContentPosition property to React Native for Web (RNW), aligning it with the recent updates in React Native (RN). The implementation is based on the changes from the following upstream PRs:

RN upstream: [facebook/react-native#35993](https://github.com/facebook/react-native/pull/35993)
RNW upstream: [necolas/react-native-web#2588](https://github.com/necolas/react-native-web/pull/2588)

The initial implementation for this feature ([Expensify/App#28793](https://github.com/Expensify/App/pull/28793)) was reverted ([Expensify/App#31158](https://github.com/Expensify/App/pull/31158)) due to a bug ([Expensify/App#31143](https://github.com/Expensify/App/issues/31143)). This PR addresses and resolves that issue.

🎉 Thanks to @janicduplessis, it was found and fixed 🎉 

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/23220
PROPOSAL: 


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
1. As an employee, create an IOU with a receipt.
2. Log in as the workspace admin.
3. Navigate to the workspace chat with the employee.
4. Access the request conversation and resolve any errors as the admin.

Testing Goals:
Verify the correctness of the scrolling direction in the chat list.
Ensure the copying process functions as expected.
Confirm that styles are properly applied (with nothing appearing inverted).
- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

perunt committed:
Merge branch 'main' of https://github.com/Expensify/App

perunt committed:
Merge branch 'main' of https://github.com/Expensify/App

perunt committed:
create MVCPFlatList

perunt committed:
use MVCPFlatList

perunt committed:
Merge branch 'perunt/chat-loader-flickering' of https://github.com/margelo/expensify-app-fork into feat/##23220-web-maintainVisibleContentPosition

perunt committed:
Merge branch 'perunt/chat-loader-flickering' of https://github.com/margelo/expensify-app-fork into feat/##23220-web-maintainVisibleContentPosition

perunt committed:
move scrollToOffset into requestAnimationFrame

perunt committed:
Merge branch 'perunt/chat-loader-flickering' of https://github.com/margelo/expensify-app-fork into feat/##23220-web-maintainVisibleContentPosition

cedarbaum commented:
Hey all,

I've been following this PR as well as https://github.com/necolas/react-native-web/pull/2588, since I would also like to get `maintainVisibleContentPosition` working in RNW.

I've got it mostly working, but wanted to point out one thing I noticed in the usage of the `MutationObserver`. If the subtree of the `FlatList` has _attributes_ that change frequently, the observe method will get called each time. I specifically saw this happening when some items in my list had repeating animations via React Native Reanimated, which caused nonstop calls to the observer.

I fixed this for my use case by removing observation of attribute changes altogether:

```ts
mutationObserver.observe(contentView, {
  childList: true,
  subtree: true,
});
```

This works for me, since I am only interested in layout changes due to items being prepended to the list. I am not confident that removing this for all cases is safe, but perhaps at least making it configurable or allowing more granular control over which specific attribute changes to look for could avoid these cases.

@janicduplessis FYI, since this could be relevant to the upstream PR.

Happy to provide more details if needed. Also, wanted to say thanks for the work on this issue and for making it open source!

cedarbaum commented:
> I did think about that, but thought it wouldn’t handle cases where items are resized. You do bring up a good point though. Thinking about it more I wonder if it would be possible to use ResizeObserver on the content view instead since we only want to maintain position when the content view size changes.

I think `ResizeObserver` could solve the issue I mentioned, though I did find another interesting case. When the content view changes height due to reasons besides items being added to the list (e.g., a cell is expanded), this can also trigger the scroll offset behavior. In such cases, it is more desirable IMO to maintain the scroll offset, since otherwise the expanded cell can shift upwards and be clipped from the top. So my new solution is to only observe the `childList` change. Using `ResizeObserver` would solve the original issue I mentioned but would not avoid this case.

However, I am not sure how `maintainVisibleContentPosition` handles such cases on other platforms. I think ideally this would be configurable in RN (i.e. allow users to specify whether they only want `maintainVisibleContentPosition` behavior due to item count changes or always).

Sorry if this is a bit off-topic from this PR, but wanted to mention in case it is relevant.

perunt committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/##23220-web-maintainVisibleContentPosition

roryabraham commented:
I'm not particularly concerned about that behavior judging by the videos. What happens if there's more than one page of messages in the chat?

perunt committed:
fix scrolling

perunt committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/##23220-web-maintainVisibleContentPosition

perunt committed:
remove scrollToOffset

perunt commented:
@0xmiroslav, could you test it again?

perunt commented:
I couldn't reproduce it in the sandbox/bare project, so I revised the scrolling functionality. I found some redundant code. Merged main. After the merge, I wasn't able to reproduce that autoscrolling anymore.

roryabraham commented:
Ok, can somebody please summarize next steps and any outstanding issues here?

janicduplessis commented:
I am able to reproduce the issue, it seems to repro best in safari, from what I understand it is related to a mix of `inverted` and `maintainVisibleContentPosition`, if I remove either then it works fine.

janicduplessis commented:
Seems also related to this patch https://github.com/Expensify/App/blob/main/patches/react-native-web%2B0.19.9%2B001%2Binitial.patch, it changes the order that items are added so I suspect this causes some problems somewhere.

janicduplessis commented:
Looks like https://github.com/Expensify/App/pull/32843 prevents the issue from happening so I guess this is fine. I think there might still be something wrong with https://github.com/Expensify/App/blob/main/patches/react-native-web%2B0.19.9%2B001%2Binitial.patch however.

janicduplessis commented:
After doing some more testing it seems like there is still an issue caused by this patch https://github.com/Expensify/App/blob/main/patches/react-native-web%2B0.19.9%2B001%2Binitial.patch, it introduces a pretty big behaviour change in how Virtualized renders its items, I am curious why it was added and what it is supposed to fix. I wonder what the best way forward is, can we revert the patch? Or should I try to figure out a fix? Would be great if someone with more context about that patch can assist.

perunt commented:
As far as I know, the patch at https://github.com/Expensify/App/blob/main/patches/react-native-web%2B0.19.9%2B001%2Binitial.patch seems to fix a selection issue. Without this patch, when you try to select text from the bottom right to the top left corner in a chat message, it ends up selecting text in each message incorrectly from left to right. This patch probably corrects this, allowing for the right to left selection. I haven't tried removing it yet, so this is just a guess.

perunt commented:
found that issue https://github.com/necolas/react-native-web/pull/2345

roryabraham commented:
If you confirm that the original bug is no longer reproducible, even with the patch removed, that's great!

roryabraham commented:
@azimgd can you speak to whether or not https://github.com/Expensify/App/blob/main/patches/react-native-web%2B0.19.9%2B001%2Binitial.patch is still necessary for some reason or if the issue was fixed in some other context?

perunt committed:
remove autoscrollToTopThreshold

perunt committed:
automatically scroll to the bottom of the chat when the component mounts

perunt committed:
fix MVCPFlatList

perunt committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/##23220-web-maintainVisibleContentPosition

perunt commented:
@0xmiroslav , can you check it again?
I’ve added scrolling to the end without animation. Since we're showing threaded reportActions in the LHN, these get fetched as well. Sometimes, this means the report might have some reportActions before openReport is triggered. 
So, with our current feature of saving the scroll position, we need to enable automatic scrolling to the end when the chat mounts. I'll work on the differentiation and additional logic for Comment Linking later.
Also, I updated the call to getContentView because on the first mount it could be zero, and we wouldn't be able to keep the viewport location unchanged.

perunt committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/##23220-web-maintainVisibleContentPosition

perunt committed:
post-merge updates

perunt commented:
@0xmiroslav have you had a chance to look into it?

janicduplessis committed:
Remove rn web invertion patch

janicduplessis commented:
I just pushed an additional commit to remove a problematic RN web patch, it should help with the issue.

perunt committed:
Merge branch 'feat/##23220-web-maintainVisibleContentPosition' of https://github.com/margelo/expensify-app-fork into feat/##23220-web-maintainVisibleContentPosition

janicduplessis committed:
Inverted fix

janicduplessis commented:
@0xmiroslav We found one last issue that I just fixed, I cannot reproduce the scrolling issue anymore. It is good to test, thanks!

janicduplessis commented on src/components/InvertedFlatList/BaseInvertedFlatList.js:
```diff
@@ -2,8 +2,6 @@ import PropTypes from 'prop-types';
 import React, {forwardRef} from 'react';
 import FlatList from '@components/FlatList';
 
-const AUTOSCROLL_TO_TOP_THRESHOLD = 128;
```
Do we actually want to remove this?

perunt replied:
`autoscrollToTopThreshold` will scroll down the list in the comment-linking process, so we need to ditch it.

janicduplessis reviewed (commented):
@perunt Are the other scrolling changes still needed after the last fixes?

perunt committed:
Merge branch 'feat/##23220-web-maintainVisibleContentPosition' of https://github.com/margelo/expensify-app-fork into feat/##23220-web-maintainVisibleContentPosition

perunt committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/##23220-web-maintainVisibleContentPosition

perunt committed:
Merge branch 'perunt/bidirectional-list-loader-mobile-fix' of https://github.com/margelo/expensify-app-fork into feat/##23220-web-maintainVisibleContentPosition

perunt commented:
@janicduplessis, it seems that due to the 'true' inversion, we're rendering from bottom to top, causing each subsequent item to overlap the one before it. Could we assign a higher zIndex to the lastFocusedCellRenderer to resolve this issue?

perunt committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/##23220-web-maintainVisibleContentPosition

perunt committed:
fix after merge

perunt committed:
use CellRendererComponent

perunt committed:
fix reassurePerformanceTests

perunt committed:
remove node_modules

perunt committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/##23220-web-maintainVisibleContentPosition

perunt committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/##23220-web-maintainVisibleContentPosition

perunt committed:
the package name is extracted from the patch file name

perunt committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/##23220-web-maintainVisibleContentPosition

roryabraham commented:
> Could we assign a higher zIndex to the lastFocusedCellRenderer to resolve this issue?

Heads-up @perunt, we have a [custom `CellRendererComponent` that applies `zIndex` sequentially by index](https://github.com/Expensify/App/blob/fc9393f67213d60b0532b73c743effeec66f20ad/src/components/InvertedFlatList/CellRendererComponent.tsx#L16-L23). We do this to make the context menu, which is a child of the list item, overlap and render on top of the next list item in the list:

<img width="489" alt="image" src="https://github.com/Expensify/App/assets/47436092/bffd14f3-09a4-4d9a-8e36-ad7a332017c7">

If they had the same zIndex, then the context menu would be cut off. Just giving a higher zIndex to the context menu doesn't work because of some nuance of stacking contexts I discovered when I first built that context menu but have since forgotten the details of 🙂

If memory serves, because the context menu is in a separate stacking context from the next item over in the list, zIndex will basically not have any effect across separate stacking contexts.

An alternate solution (which might be good for performance, too), might be to use a portal and essentially have only one absolute-positioned `MiniReportActionContextMenu` and update the position based on which `ReportActionItem` is hovered. Rather than having a separate `MiniReportActionContextMenu` for every single item in the list. As long as the position updates fast enough that we don't see jankiness or other weird bugs, that could be a good change. If we're going down that route, it could also be kind of cool to create an animation that translates the `MiniReportActionContextMenu` position.

If you think that's a good idea, let's be sure to talk about it a bit in slack before you start implementing it.

perunt commented:
We've aligned our web setup with the mobile version by using CellRendererComponent for now. I totally understand the hassle of tweaking zIndex, so the idea of switching to Portals is quite appealing. Last week's test revealed that the main delay in our current setup is the context menu's slow initial appearance. We might also consider a similar approach as used in the Tooltip component for consistency.

As of now, things are functioning as they were. Currently, I'm focused on failed tests.
Transitioning to a Portal-based context menu seems like a smart strategy for enhanced performance going forward. How about we start integrating this idea in our upcoming PRs? Of course, let's first discuss it in the channel to get everyone's input.

perunt committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/##23220-web-maintainVisibleContentPosition

perunt committed:
cleaning up reassurePerformanceTests

perunt committed:
cache clean

perunt committed:
patch-package

perunt committed:
move reportScrollManager from ReportActionsView to ReportActionsList to

perunt committed:
do patch-package again

perunt committed:
postinstall update

perunt committed:
undo postinstall

perunt committed:
Backup current patches

perunt committed:
indentation

perunt committed:
undo reassurePerformanceTests

perunt commented:
Hi @0xmiroslav, I found that our `Performance Tests` failing issue is due to applying patches from the main branch to our feature branch. You can see the specific workflow step here: [GitHub Workflow File](https://github.com/margelo/expensify-app-fork/blob/454f8a13a304b8cc51f9149253cc83c696fad003/.github/workflows/reassurePerformanceTests.yml#L42).

Once we merge this PR, the source of our patches will be correctly aligned with our branch, which should resolve the test failures. However, if this approach fails to address the issue thoroughly, I'll do a fix in a separate PR. 
Would that work for you?

perunt committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/##23220-web-maintainVisibleContentPosition

perunt committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/##23220-web-maintainVisibleContentPosition

perunt committed:
cleanup reassurePerformanceTests

perunt commented on .github/workflows/reassurePerformanceTests.yml:
```diff
@@ -21,7 +21,6 @@ jobs:
         run: |
             git config --global user.email "test@test.com"
             git config --global user.name "Test"
-
```
👍🏻

perunt committed:
lint

perunt committed:
add spaces after linting

perunt commented:
Done.
I see the[ issue with the linter ](https://expensify.slack.com/archives/C01GTK53T8Q/p1704387355957189?thread_ts=1704302068.776969&cid=C01GTK53T8Q) so it's failing due to that

perunt committed:
Merge branch 'main' of https://github.com/Expensify/App into feat/##23220-web-maintainVisibleContentPosition

perunt commented:
done
lint is green

perunt commented:
@0xmiroslav, is it good for you? Can we get it merged?

roryabraham commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/47436092/a0eca122-6c1f-4f4a-82e1-32ebd515269f

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

roryabraham commented:
Note that I was able to reproduce the weird scrolling issue, but only on Android Chrome, which on a dev environment in an emulator was super janky anyways. So I'm not too concerned by that and sort of doubt we'll see it in production.

roryabraham commented on src/components/FlatList/MVCPFlatList.js:
```diff
@@ -0,0 +1,207 @@
+/* eslint-disable es/no-optional-chaining, es/no-nullish-coalescing-operators, react/prop-types */
```
Note for posterity: these were disabled to keep this code as similar as possible to the upstream PR it's based on: https://github.com/necolas/react-native-web/pull/2588

Hopefully this component only lives in our codebase relatively temporarily. We probably will want to switch to FlashList soon too.

roryabraham commented:
Merging through failing reassure per [this comment](https://github.com/Expensify/App/pull/32098#issuecomment-1874157035)

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

janicduplessis commented:
Maybe this should not have been removed? https://github.com/Expensify/App/pull/32098/files#diff-b1c8c4fd8072edbd35c63a7cdd71b63ff1701a6f43529ea794f7720a86460df2L18

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7470902953) to staging by https://github.com/roryabraham in version: 1.4.24-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

shubham1206agra commented:
> chat is not scrolled down automatically to see last report action after request money in workspace chat. Need to confirm if this is caused by this PR

```js
useEffect(() => {
        if (previousLastIndex.current !== lastActionIndex && reportActionSize.current > sortedVisibleReportActions.length) {
            reportScrollManager.scrollToBottom();
        }
        previousLastIndex.current = lastActionIndex;
        reportActionSize.current = sortedVisibleReportActions.length;
    }, [lastActionIndex, sortedVisibleReportActions.length, reportScrollManager]);
```
Maybe update this logic to fix this issue.

perunt commented:
Going to check this out. Might fix it by going back to [https://github.com/Expensify/App/pull/32098/files#diff-b1c8c4fd8072edbd35c63a7cdd71b63ff1701a6f43529ea794f7720a86460df2L18], like @janicduplessis said. But, this could mess up comment linking in the next PR. Still, it's a decent quick fix for staging.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7486260559) to production by https://github.com/thienlnam in version: 1.4.24-3 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

s77rt commented on src/components/FlatList/MVCPFlatList.js:
```diff
@@ -0,0 +1,207 @@
+                scrollToOffset(0, true);
+            }
+        }
+    }, [getScrollOffset, scrollToOffset, mvcpMinIndexForVisible, mvcpAutoscrollToTopThreshold, horizontal]);
```
Having `mvcpAutoscrollToTopThreshold` as a dependency made this function gets re-evaluated on value change. Which also made the `setupMutationObserver` function re-evaluated and executed (due to `useEffect`). That later function will disconnect the old observer and creates a new one. This constant observer creation is found to disrupt the new message detection.

alitoshmatov commented on src/components/FlatList/MVCPFlatList.js:
```diff
@@ -0,0 +1,207 @@
+
+        mutationObserverRef.current?.disconnect();
+
+        const mutationObserver = new MutationObserver(() => {
```
When list is hidden and reappears, it scrolls down to the bottom. But it should keep its last scroll offset. That is why ignore this callback when list is hidden. Coming from https://github.com/Expensify/App/issues/45434

blazejkustra commented on patches/react-native-web+0.19.9+001+initial.patch:
```diff
@@ -1,286 +1,648 @@
-  *
-  */
- class VirtualizedList extends StateSafePureComponent {
-+  pushOrUnshift(input, item) {
```
Commenting to cross-reference this in another issue. I believe this PR caused [this](https://github.com/Expensify/App/issues/37447)
